### PR TITLE
Fix for C SDK Issue segfault when using wolfSSL and MQTT/WS

### PR DESCRIPTION
**Pull request by momuno**
This is a fix for C SDK Issue #1759 [https://github.com/Azure/azure-iot-sdk-c/issues/1759].

It involves two fixes:
1) Addresses the segfault using wolfSSL and MQTT/WS.  A callback was not reset to NULL and was mistakenly being called resulting in the segfault. (It should only be called on application data).
2) Updates logic to not return WANT_READ during renegotiation when waiting to receive CONNACK.

This change requires https://github.com/wolfSSL/wolfssl/pull/3594 to be merged, and the following wolfSSL release to be used. (v4.7.0 to be released mid-February)
Assuming this has occurred, wolfSSL must be configured with --enable-secure-renegotiation.

Enabling secure renegotiation does not obstruct using MQTT or AMQP.  It DOES enable MQTT/WS and AMQP/WS.

### CODE REVIEW

**Review by dgarske — APPROVED** [review]
@momuno  These changes look good.

File: adapters/tlsio_wolfssl.c [hunk 1]
```diff
@@ -389,7 +389,8 @@ static int on_io_recv(WOLFSSL *ssl, char *buf, int sz, void *context)
         while (tls_io_instance->socket_io_read_byte_count == 0 && socket_reads < SOCKET_READ_LIMIT)
         {
             xio_dowork(tls_io_instance->socket_io);
-            if (tls_io_instance->tlsio_state != TLSIO_STATE_IN_HANDSHAKE)
+            if (tls_io_instance->tlsio_state != TLSIO_STATE_IN_HANDSHAKE &&
+                tls_io_instance->tlsio_state != TLSIO_STATE_OPEN) // Renegotiation
             {
                 break;
```
Comment (danewalton) [inline adapters/tlsio_wolfssl.c:395]: `LogInfo` here?

File: adapters/tlsio_wolfssl.c [hunk 2]
```diff
@@ -562,6 +565,14 @@ static int create_wolfssl_instance(TLS_IO_INSTANCE* tls_io_instance)
         wolfSSL_SetIOWriteCtx(tls_io_instance->ssl, tls_io_instance);
         wolfSSL_SetIOReadCtx(tls_io_instance->ssl, tls_io_instance);
 
+#ifdef HAVE_SECURE_RENEGOTIATION
+        if(wolfSSL_UseSecureRenegotiation(tls_io_instance->ssl) != SSL_SUCCESS)
+        {
+            LogError("unable to enable secure renegotiation");
+            return MU_FAILURE;
```
Comment (danewalton) [inline adapters/tlsio_wolfssl.c:572]: I believe the code style in this repo is to have one return point except parameter checking. This should likely be
```suggestion
            result = MU_FAILURE;
```
and may have to move around to work with the other `if/else if's`

**Review by danewalton — APPROVED** [review]
Please update that result to not return in the middle of the function. Other than that lgtm

File: adapters/tlsio_wolfssl.c (hunk 2 above)
Comment (momuno) [reply adapters/tlsio_wolfssl.c:572]: In this case, if result = MU_FAILURE, result will get reset to 0 at line 577.  So it should be returned here.

File: adapters/tlsio_wolfssl.c [hunk 3]
```diff
@@ -564,6 +565,14 @@ static int create_wolfssl_instance(TLS_IO_INSTANCE* tls_io_instance)
 
         tls_io_instance->tlsio_state = TLSIO_STATE_NOT_OPEN;
         result = 0;
+
+#ifdef HAVE_SECURE_RENEGOTIATION
+        if(wolfSSL_UseSecureRenegotiation(tls_io_instance->ssl) != SSL_SUCCESS)
```
Comment (ericwol-msft) [inline adapters/tlsio_wolfssl.c:570]:
```suggestion
        if (wolfSSL_UseSecureRenegotiation(tls_io_instance->ssl) != SSL_SUCCESS)
```

File: adapters/tlsio_wolfssl.c [hunk 4]
```diff
@@ -892,6 +901,10 @@ int tlsio_wolfssl_send(CONCRETE_IO_HANDLE tls_io, const void* buffer, size_t siz
             {
                 result = 0;
             }
+
+            // remove on send complete and callback context
+            memset((void*)&tls_io_instance->on_send_complete, 0, sizeof(tls_io_instance->on_send_complete));
```
Comment (ericwol-msft) [inline adapters/tlsio_wolfssl.c:906]: isn't this just setting the pointer to NULL?

**Review by ericwol-msft — APPROVED** [review]

File: adapters/tlsio_wolfssl.c (hunk 4 above)
Comment (momuno) [reply adapters/tlsio_wolfssl.c:906]: It is, but I wanted to mimic the code that was already in tlsio_mbedtls.c doing its own reset (albeit of a struct containing the function pointer and the callback context).  Maybe it's not worth it.

**Pull request merged**